nixos/modules/services/misc/matrix-dendrite.nix

```diff
@@ -0,0 +1,342 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.matrix-dendrite;
+  settingsFormat = pkgs.formats.yaml {};
+  configurationYaml = settingsFormat.generate "dendrite.yaml" cfg.settings;
+in {
+  options.services.matrix-dendrite = {
+      enable = mkEnableOption "matrix.org dendrite";
+      dataDir = mkOption {
+        type = types.str;
+        default = "/var/lib/matrix-dendrite";
+        description = ''
+          The directory where matrix-dendrite stores its stateful data such as
+          certificates, media and uploads.
+        '';
+      };
+      settings = lib.mkOption {
+        type = lib.types.submodule {
+          freeformType = settingsFormat.type;
```

Comment:
Not that it matters very much, but I'm seeing this a lot. Freeform modules aren't needed if the submodule doesn't declare any options. In this case this is equivalent to just
```nix
{
  type = settingsFormat.type;
}
```

Reply:
Out of curiosity, when is it better to declare options under submodules vs options under the toplevel  `options.services.matrix-dendrite` ? I want to understand this better, and hopefully contribute a bit to some tutorials and documentation.

Reply:
And upon re-reading everything and realizing that options is a required attribute of types.submodule, everything works! Thanks!

Reply:
For `settings`, a submodule is used to delimit where values can be freely set using `freeformType`. Within a `freeformType` submodule, you can declare values without having associated options. Whereas outside the submodule, every value needs to have an associated option (unless the parent module also has a `freeformType` set).

Reply:
Thank you! That makes sense. One other question: Settings itself is ultimately written (in it's entirety) as the configuration yaml (dendrite.yaml). There are probably other settings that I would want to declare without associated options which would be used to populate the command line flags to the daemon, which I'm currently doing with options (since if I put these under settings, they will be incorrectly written to the yaml). Is there a convention for multiple submodules that you had in mind, when you have multiple different configuration files for example?